numpy/lib/shape_base.py

```diff
@@ -89,9 +90,20 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     outshape = asarray(arr.shape).take(indlist)
     i.put(indlist, ind)
     res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
-    #  if res is a number, then we have a smaller output array
+
+    # see gh-8363
+    _strcase = False
+    _dtypes = set()
+
+    # If res is a number, then we have a smaller output array.
     if isscalar(res):
```

Comment:
~~~python
np.array([1, 2], dtype=object).view(np.int8)
~~~
> This should be performant as the solution you were proposing, I'd hope
Can you elaborate on this?  As you saw, I don't really understand what @shoyer is talking about (but you seem to), so some elucidation would be good here.